lib/service/awsoidc.go

```diff
@@ -0,0 +1,330 @@
+	// an RDS database.
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
+		if database.IsAWSHosted() && database.IsRDS() {
+			awsRegions[database.GetAWS().Region] = nil
+		}
+	}
+
+	integrations, err := updater.AuthClient.ListAllIntegrations(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Perform updates in parallel across regions.
+	var sem = make(chan interface{}, maxConcurrentUpdates)
+	var wg sync.WaitGroup
+	for _, ig := range integrations {
+		for region := range awsRegions {
+			sem <- nil
+			wg.Add(1)
+			go func(ig types.Integration, region string) {
+				if err := updater.updateDeployServiceAgent(ctx, ig, region, stableVersion); err != nil {
+					updater.Log.WithError(err).Warning("Failed to update deploy service agent.")
+				}
+				wg.Done()
+				<-sem
+			}(ig, region)
+		}
+	}
+	wg.Wait()
+
+	return nil
+}
+
+func (updater *DeployServiceUpdater) updateDeployServiceAgent(ctx context.Context, integration types.Integration, awsRegion, teleportVersion string) error {
+	// Do not attempt update if integration is not an aws oidc integration.
+	if integration.GetAWSOIDCIntegrationSpec() == nil {
+		return nil
+	}
+
+	token, err := updater.AuthClient.GenerateAWSOIDCToken(ctx, types.GenerateAWSOIDCTokenRequest{
+		Issuer: updater.AWSOIDCProviderAddr,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	req := &awsoidc.AWSClientRequest{
+		IntegrationName: integration.GetName(),
+		Token:           token,
+		RoleARN:         integration.GetAWSOIDCIntegrationSpec().RoleARN,
+		Region:          awsRegion,
+	}
+
+	// The deploy service client is initialized using AWS OIDC integration.
+	deployServiceClient, err := awsoidc.NewDeployServiceClient(ctx, req, updater.AuthClient)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// ownershipTags are used to identify if the ecs resources are managed by the
+	// teleport integration.
+	ownershipTags := map[string]string{
+		types.ClusterLabel:     updater.TeleportClusterName,
+		types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+		types.IntegrationLabel: integration.GetName(),
+	}
+
+	// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+	// If the lease cannot be acquired, the update is already being handled by another instance.
+	semLock, err := updater.AuthClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+		SemaphoreKind: types.SemaphoreKindConnection,
+		SemaphoreName: fmt.Sprintf("update_deploy_service_agents_%s_%s", awsRegion, integration.GetName()),
+		MaxLeases:     1,
+		Expires:       updater.Clock.Now().Add(updateDeployAgentsInterval),
+		Holder:        "update_deploy_service_agents",
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), teleport.MaxLeases) {
+			updater.Log.WithError(err).Debug("Deploy service agent update is already being processed.")
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+
```

Comment:
Is there a reason why we only call the `CancelSemaphoreLease` on error?

I would expect that we would call `defer ... CancelSemaphoreLease` here

Reply:
If we hold on to the semaphore lease, other instances will not be able to acquire the lease within the same 30min interval. This will reduce the number of API calls a bit.

A small optimization, but could potentially be prone to bugs in future code changes. I'll go ahead and just cancel the semaphore on success as well.